domain_config/src/DomainConfigServiceProvider.php

```diff
@@ -0,0 +1,25 @@
+<?php
+
+/**
+ * @file
+ * Contains \Drupal\domain_config\DomainConfigServiceProvider.
+ */
+
+namespace Drupal\domain_config;
+
+use Drupal\Core\DependencyInjection\ContainerBuilder;
+use Drupal\Core\DependencyInjection\ServiceProviderBase;
+
+/**
+ * Overrides the router.route_provider service to point to our customized one.
+ */
+class DomainConfigServiceProvider extends ServiceProviderBase {
+
+  /**
+   * {@inheritdoc}
+   */
+  public function alter(ContainerBuilder $container) {
+    $definition = $container->getDefinition('router.route_provider');
+    $definition->setClass('Drupal\domain_config\Routing\DomainRouteProvider');
```

Comment:
In PHP 5.5, it's safe and I would argue preferred to use the class-constant-form for a class name.  That is:

```php
setClass(DomainRouteProvider::class);
```

And then also add a "use" statement above for that class. That way your IDE can recognize it as a class name, spell check it, find it in refactor commands, etc. etc.